Refactor `move_constructor_to_cuda`.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114626

Follow-up: #114539

This PR introduces a minor change to the `move_constructor_to_cuda` implementation, while
refactoring the whole pass into a class. Here's a brief summary of the changes:

- Create a new `ConstructorMoverPass`
- Rephrase the condition:

```python
if not isinstance(
    node.target, torch._ops.OpOverload
) or node.target.namespace not in ("prims", "aten"):
    ...

if not (
    isinstance(node.target, torch._ops.OpOverload)
    and node.target.namespace in ("prims", "aten")
):
    ...
```

cc @bdhirsh @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

ysiraichi committed:
Refactor `move_constructor_to_cuda`.

Follow-up: #114539

This PR introduces a few minor changes to the `move_constructor_to_cuda`
implementation. Here's a brief summary of the changes:

- Use `defaultdict` instead of `Counter`
- Raise `ZeroOrMultipleDevicesError` if we find anything other than a single CUDA device
    - We catch that error, and pass
- Rephrase the condition:

```python
if not isinstance(
    node.target, torch._ops.OpOverload
) or node.target.namespace not in ("prims", "aten"):
    ...

if not (
    isinstance(node.target, torch._ops.OpOverload)
    and node.target.namespace in ("prims", "aten")
):
    ...
```

[ghstack-poisoned]

ysiraichi committed:
Update on "Refactor `move_constructor_to_cuda`."


Follow-up: #114539

This PR introduces a few minor changes to the `move_constructor_to_cuda`
implementation. Here's a brief summary of the changes:

- Use `defaultdict` instead of `Counter`
- Raise `ZeroOrMultipleDevicesError` if we find anything other than a single CUDA device
    - We catch that error, and pass
- Rephrase the condition:

```python
if not isinstance(
    node.target, torch._ops.OpOverload
) or node.target.namespace not in ("prims", "aten"):
    ...

if not (
    isinstance(node.target, torch._ops.OpOverload)
    and node.target.namespace in ("prims", "aten")
):
    ...
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ysiraichi committed:
Rebased on "Refactor `move_constructor_to_cuda`."


Follow-up: #114539

This PR introduces a few minor changes to the `move_constructor_to_cuda`
implementation. Here's a brief summary of the changes:

- Use `defaultdict` instead of `Counter`
- Raise `ZeroOrMultipleDevicesError` if we find anything other than a single CUDA device
    - We catch that error, and pass
- Rephrase the condition:

```python
if not isinstance(
    node.target, torch._ops.OpOverload
) or node.target.namespace not in ("prims", "aten"):
    ...

if not (
    isinstance(node.target, torch._ops.OpOverload)
    and node.target.namespace in ("prims", "aten")
):
    ...
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

JackCaoG commented:
@eellison We would like to have this pr(and the https://github.com/pytorch/pytorch/pull/114627) to be merged before the branch cut(Dec 1, this Friday). Do you think this is possible? This is an important pr since it will unblock a couple torchbench model on our end.

JackCaoG commented:
@eellison Thanks a lot!

eellison commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -95,7 +95,7 @@ def post_grad_passes(gm: torch.fx.GraphModule, is_inference: bool):
     stable_topological_sort(gm.graph)
 
-    move_constructors_to_cuda(gm.graph)
+    move_constructors_to_cuda(gm)
```
Can we switch this back to taking in a graph ? We don't want to do a gm recompile here, and I think taking in a graph instead of a gm makes it more clear that this pass does not recompile.

ysiraichi replied:
Yes. All FX passes take in a GraphModule.

ysiraichi replied:
But, since we are not moving this pass to torch/fx/passes, we can simply not inherit from `PassBase`, and take in a graph.

eellison commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -1070,19 +1079,29 @@ def move_constructors_to_cuda(graph):
+        return
+
+    if len(cuda_devices) != 1:
+        raise ZeroOrMultipleDevicesError("cuda", cuda_devices)
```
This pass shouldn't error, it's just a missed optimization if there are multiple devices.  Can we remove this ? 

In general, this pass has no guarantee of moving all compute away from cpu. If there are cpu inputs or outputs, or non movable operators, this won't succed.

The xla pass can do a follow up pass to see if it was successful.

eellison commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -1070,19 +1079,29 @@ def move_constructors_to_cuda(graph):
 
 
-def find_movable_constructors(graph, constructors: List[fx.Node]) -> Set[fx.Node]:
+def find_movable_constructors(graph, constructors: List[fx.Node]) -> List[fx.Node]:
```
Can we keep this as a set? it makes more clear that the output entries will be unique

eellison commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -1131,6 +1150,10 @@ def make_dependencies_equivalent(
                 and node_device
                 and node_device.type == "cuda"
             ):
+                # cpu_indeg is a defaultdict.
+                # The line below is needed so as to create, if not existent, the entry in
+                # the dictionary. Then, we can delete it. Otherwise, we get a KeyError.
+                cpu_indeg[user]
```
If we just keep this as a Counter we don't need these changes. can we go back to Counter ?

eellison commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -1152,4 +1175,18 @@ def make_dependencies_equivalent(
+
+    try:
+        move_constructors_to_cuda_impl(graph_module)
+    except ZeroOrMultipleDevicesError:
```
I personally prefer not to use Exceptions as control flow.

eellison reviewed (commented):
have some comments. I would prefer not moving this to general torch/fx pass because this pass only works on aten-fx ir. I'm not sure where the best place to expose it. Maybe we could have a public torch/inductor/utils directory. I'm not sure exactly. 

thoughts @JackCaoG ?

ysiraichi commented:
Then, what do you think about leaving it in torch/_inductor/fx_passes/post_grad.py? We can move it afterwards, if we think it's necessary.

ysiraichi committed:
Update on "Refactor `move_constructor_to_cuda`."


Follow-up: #114539

This PR introduces a few minor changes to the `move_constructor_to_cuda`
implementation. Here's a brief summary of the changes:

- Use `defaultdict` instead of `Counter`
- Raise `ZeroOrMultipleDevicesError` if we find anything other than a single CUDA device
    - We catch that error, and pass
- Rephrase the condition:

```python
if not isinstance(
    node.target, torch._ops.OpOverload
) or node.target.namespace not in ("prims", "aten"):
    ...

if not (
    isinstance(node.target, torch._ops.OpOverload)
    and node.target.namespace in ("prims", "aten")
):
    ...
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ysiraichi committed:
Update on "Refactor `move_constructor_to_cuda`."


Follow-up: #114539

This PR introduces a minor change to the `move_constructor_to_cuda` implementation, while
refactoring the whole pass into a class. Here's a brief summary of the changes:

- Create a new `ConstructorMoverPass`
- Rephrase the condition:

```python
if not isinstance(
    node.target, torch._ops.OpOverload
) or node.target.namespace not in ("prims", "aten"):
    ...

if not (
    isinstance(node.target, torch._ops.OpOverload)
    and node.target.namespace in ("prims", "aten")
):
    ...
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ysiraichi commented:
@eellison I have addressed your suggestions. Could you take a look at it again?

eellison commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -987,169 +987,203 @@ def fused_int_mm_mul(match: Match, mat1, mat2, mat3, out_dtype=None):
-    if node.target == "output":
-        return True
+        A constructor node can be moved to the target_device iff all of its users
+        can also be moved (tested by cannot_be_moved). Otherwise, all dependant
```
nit: `s/dependant/dependent`

eellison reviewed (approved):
Looks good, thanks!

ysiraichi committed:
Update on "Refactor `move_constructor_to_cuda`."


Follow-up: #114539

This PR introduces a minor change to the `move_constructor_to_cuda` implementation, while
refactoring the whole pass into a class. Here's a brief summary of the changes:

- Create a new `ConstructorMoverPass`
- Rephrase the condition:

```python
if not isinstance(
    node.target, torch._ops.OpOverload
) or node.target.namespace not in ("prims", "aten"):
    ...

if not (
    isinstance(node.target, torch._ops.OpOverload)
    and node.target.namespace in ("prims", "aten")
):
    ...
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

atalman commented:
@ysiraichi could you please take a look at the failure:
[inductor / linux-jammy-cpu-py3.8-gcc11-inductor / test (cpu_inductor_torchbench, 1, 2, linux.4xlarge)](https://hud.pytorch.org/pr/pytorch/pytorch/114626#19154246330) ([gh](https://github.com/pytorch/pytorch/actions/runs/7037896475/job/19154246330))

ysiraichi commented:
@pytorchbot merge

ysiraichi commented:
@pytorchbot merge
